### PR TITLE
Adding support for re2 and hyperscan for like operator

### PR DESCRIPTION
**Pull request by grajoria**
Signed-off-by: Girjesh Rajoria <girjesh.rajoria@gmail.com>

### CODE REVIEW

File: include/s3select_functions.h [hunk 1]
```diff
@@ -1777,23 +1776,26 @@ bool operator()(bs_stmt_vec_t* args, variable* result) override
...
     value main_expr_val = main_expr->eval();
     if (main_expr_val.type != value::value_En_t::STRING)  {
         throw base_s3select_exception("main expression must be string");
     }
-         
-    if ( std::regex_match(main_expr_val.to_string(), compiled_regex)) {
+
+    std::string content = main_expr_val.to_string();
+    std::string like_as_regex_str(like_as_regex.begin(), like_as_regex.end());
+    RE2 compiled_regex(like_as_regex_str);
```
Comment (galsalomon66) [inline include/s3select_functions.h:1789]: `compiled_regex` should be a class member 
upon constant_state is true it `compiled_regex` is compiled once instead of many times.

**Comment by galsalomon66** [comment]
i left a comment.

since the RE2 code is quite similar to the previous one, it needs to add more abstraction, to make it easier to introduce other regular-expression engines.

the match/and compile functionalities should be abstract.

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,7 +2,7 @@ cmake_minimum_required(VERSION 3.0)
 
 project(s3select)
 
-set(CMAKE_CXX_FLAGS "-std=gnu++17 -ggdb -Wnon-virtual-dtor -Wreorder -Wunused-variable -Wtype-limits -Wsign-compare -Wmaybe-uninitialized")
+set(CMAKE_CXX_FLAGS "-std=gnu++17 -ggdb -Wnon-virtual-dtor -Wreorder -Wunused-variable -Wtype-limits -Wsign-compare -Wmaybe-uninitialized -lre2")
```
Comment (galsalomon66) [inline CMakeLists.txt:5]: need to check the re2 change in a docker file (s3select Dockerfile)
is it able to build and run tests?

File: include/s3select_functions.h [hunk 3]
```diff
@@ -1518,19 +1518,10 @@ struct _fn_in : public base_function
   }
 };
 
-struct _fn_like : public base_function
+struct _fn_like : public base_like
```
Comment (galsalomon66) [inline include/s3select_functions.h:1521]: :+1:

File: include/s3select_functions.h [hunk 4]
```diff
@@ -1545,217 +1536,15 @@ struct _fn_like : public base_function
...
-
-      if (escape_expr_val.type != value::value_En_t::STRING)  {
-        throw base_s3select_exception("escape expression must be string");
-      }
+      param_validation(esc, like_expr);
 
       std::vector<char> like_as_regex = transform(like_expr_val.str(), *escape_expr_val.str());
       std::string like_as_regex_str(like_as_regex.begin(), like_as_regex.end());
-      compiled_regex = std::regex(like_as_regex_str);
+      compiled_regex.reset(new RE2(like_as_regex_str));
```
Comment (galsalomon66) [inline include/s3select_functions.h:1543]: please use `std::make_unique` (instead of new)

File: include/s3select_functions.h [hunk 5]
```diff
@@ -1765,35 +1554,31 @@ bool operator()(bs_stmt_vec_t* args, variable* result) override
...
-        throw base_s3select_exception("esacpe expression must be string");
-      }
+    if (constant_state == false)
+    {
+      param_validation(escape_expr, like_expr);
 
       std::vector<char> like_as_regex = transform(like_expr_val.str(), *escape_expr_val.str());
       std::string like_as_regex_str(like_as_regex.begin(), like_as_regex.end());
-      compiled_regex = std::regex(like_as_regex_str);
+      compiled_regex.reset(new RE2(like_as_regex_str));
```
Comment (galsalomon66) [inline include/s3select_functions.h:1563]: please use std::make_unique (instead of new)

File: test/s3select_test.cpp [hunk 6]
```diff
@@ -2660,7 +2660,7 @@ void generate_csv_multirow(std::string& out) {
   ss << "3,35581,9091,2.1,Bibby,Primalia,Bibby.Primalia@yopmail.com,doctor,2001-02-27T23:18:23.446633-12:00" << std::endl;
   ss << "4,38388,7345,4.7,Damaris,Arley,Damaris.Arley@yopmail.com,firefighter,1995-08-24T01:40:00+12:30" << std::endl;
   ss << "5,42802,6464,7.0,Georgina,Georas,Georgina.Georas@yopmail.com,worker,2013-01-30T05:27:59.2Z" << std::endl;
-  ss << "6,45582,5863,0.1,Kelly,Hamil,Kelly.Hamil@yopmail.com,police officer,1998-03-31T17:25-01:05" << std::endl;
+  ss << "6,45582,52863,0.1,Kelly,Hamil,Kelly.Hamil@yopmail.com,police officer,1998-03-31T17:25-01:05" << std::endl;
```
Comment (galsalomon66) [inline test/s3select_test.cpp:2663]: what is the purpose of this change?

**Comment by galsalomon66** [comment]
the purpose of this change is performance (CPU perspective).
we need to verify that (RE2 vs std::regex)

File: test/s3select_test.cpp (hunk 6 above)
Comment (grajoria) [reply test/s3select_test.cpp:2663]: It is check proper anchoring done while matching compiled regex to main_expr

File: include/s3select_functions.h (hunk 4 above)
Comment (grajoria) [reply include/s3select_functions.h]: Done

File: include/s3select_functions.h (hunk 5 above)
Comment (grajoria) [reply include/s3select_functions.h]: Done

**Comment by grajoria** [comment]
@mattbenjamin please review this

File: include/s3select_functions.h [hunk 7]
```diff
@@ -1765,35 +1554,31 @@ bool operator()(bs_stmt_vec_t* args, variable* result) override
...
-        throw base_s3select_exception("main expression must be string");
+    if (main_expr_val.type != value::value_En_t::STRING)
+    {
+      throw base_s3select_exception("main expression must be string");
     }
-         
-    if ( std::regex_match(main_expr_val.to_string(), compiled_regex)) {
+
+    std::string content = main_expr_val.to_string();
+    re2::StringPiece res[1];
```
Comment (mattbenjamin) [inline include/s3select_functions.h:1573]: this exposes re2 in the header;  for functions that are large/potentially not very fast, maybe better not to inline them--but not sure

**Review by mattbenjamin — COMMENTED** [review]
agree with the top-level comment by gal, we need a strategy to abstract regex functionality, without too much loss of efficiency;  it seems like the first thing to do is look at whether we can hide the regex strategy in a templatization (for parts that we -do- want exposed in headers), if not in cpp/implementation files;  we don't want to do more allocations/deallocations in any fast path to get it, I don't think (no new new-delete of a "regex-impl"--at least, unless we can do it (e.g.) one time at startup, and then pick?

File: include/s3select_oper.h [hunk 8]
```diff
@@ -2527,6 +2528,309 @@ class base_timestamp_to_string : public base_function
...
+class base_like : public base_function
+{
+  protected:
+    value like_expr_val;
+    value escape_expr_val;
+    bool constant_state = false;
+    std::unique_ptr<RE2> compiled_regex;
+
+  public:
+    void param_validation(base_statement* escape_expr, base_statement* like_expr)
```
Comment (galsalomon66) [inline include/s3select_oper.h:2540]: :+1: :

File: include/s3select_oper.h [hunk 9]
```diff
@@ -2829,6 +2838,24 @@ class base_like : public base_function
       return v;
     }
 
+    void compile(std::vector<char>& like_regex)
+    {
+      std::string like_as_regex_str(like_regex.begin(), like_regex.end());
```
Comment (galsalomon66) [inline include/s3select_oper.h:2843]: :+1: 

should do the same for the match

File: include/s3select_oper.h [hunk 10]
```diff
@@ -2534,7 +2538,12 @@ class base_like : public base_function
     value like_expr_val;
     value escape_expr_val;
     bool constant_state = false;
-    std::unique_ptr<RE2> compiled_regex;
+    #if REGEX_HS
+      hs_database_t* compiled_regex;
```
Comment (galsalomon66) [inline include/s3select_oper.h:2542]: since the semantic is the same and the difference is type (operation on type)
maybe template could be more appropriate

File: include/s3select_functions.h [hunk 11]
```diff
@@ -1569,14 +1578,38 @@ struct _fn_like : public base_like
...
 
-    std::string content = main_expr_val.to_string();
-    re2::StringPiece res[1];
+    //matching
+    #if REGEX_HS
+      const char* content = main_expr_val.to_string().c_str();
+      res = false;
+
+      hs_scratch_t *scratch = NULL;
+      if (hs_alloc_scratch(compiled_regex, &scratch) != HS_SUCCESS)
```
Comment (galsalomon66) [inline include/s3select_functions.h:1587]: is it necessary to call alloc per each match?
once should be enough.

File: include/s3select_functions.h [hunk 12]
```diff
@@ -1569,14 +1578,38 @@ struct _fn_like : public base_like
       throw base_s3select_exception("main expression must be string");
     }
 
-    std::string content = main_expr_val.to_string();
-    re2::StringPiece res[1];
+    //matching
+    #if REGEX_HS
+      const char* content = main_expr_val.to_string().c_str();
```
Comment (galsalomon66) [inline include/s3select_functions.h:1583]: it's better to use match-function to hide this 
or better to use a template.

**Comment by galsalomon66** [comment]
I left some comments.
what is the procedure for installing hyper-scan?

**Comment by grajoria** [comment]
you can simply do - dnf install hyperscan-devel
this code should work with it

File: CMakeLists.txt [hunk 13]
```diff
@@ -13,6 +13,11 @@ if(DEFINED ENV{NANO_SEC})
   add_definitions(-DBOOST_DATE_TIME_POSIX_TIME_STD_CONFIG)
 endif()
 
+option(HYPERSCAN "For Hyperscan regex engine" OFF)
+if(HYPERSCAN)
```
Comment (galsalomon66) [inline CMakeLists.txt:17]: please review the master/CMake.
it handles the arrow library, whether it exists or not (package is installed/no-installed)

File: CMakeLists.txt [hunk 14]
```diff
@@ -9,7 +9,7 @@ if(Arrow_FOUND)
   add_definitions(-D_ARROW_EXIST)
 endif()
 
-set(CMAKE_CXX_FLAGS "-std=gnu++17 -ggdb -Wnon-virtual-dtor -Wreorder -Wunused-variable -Wtype-limits -Wsign-compare -Wmaybe-uninitialized")
+set(CMAKE_CXX_FLAGS "-std=gnu++17 -ggdb -Wnon-virtual-dtor -Wreorder -Wunused-variable -Wtype-limits -Wsign-compare -Wmaybe-uninitialized -lre2 -lhs")
```
Comment (galsalomon66) [inline CMakeLists.txt:12]: the linkage (-lre2 -lhs) should be set only whether the package is found.
find_package( ... ) --> if ( ... found ) --> add_defintions( ... )

File: include/s3select_oper.h [hunk 15]
```diff
@@ -2676,6 +2684,387 @@ class base_timestamp_to_string : public base_function
...
+        }
+      #elif REGEX_RE2
+        compiled_regex = std::make_unique<RE2>(like_as_regex_str);
+      #else
+        compiled_regex = std::regex(like_as_regex_str);
+      #endif
+    }
+
+    void match(value& main_expr_val, variable* result)
+    {
```
Comment (galsalomon66) [inline include/s3select_oper.h:3023]: :+1:

File: include/s3select_oper.h [hunk 16]
```diff
@@ -2676,6 +2684,387 @@ class base_timestamp_to_string : public base_function
...
+      }
+      return v;
+    }
+
+    void compile(std::vector<char>& like_regex)
+    {
+      std::string like_as_regex_str(like_regex.begin(), like_regex.end());
+
+      #if REGEX_HS
+	std::string temp = "^" + like_as_regex_str + "\\z";  //for anchoring start and end
```
Comment (galsalomon66) [inline include/s3select_oper.h:3002]: is that necessary? would it create conflict ?

Comment (grajoria) [reply include/s3select_oper.h:3002]: ^ and \z are for anchoring start and end of string in hyperscan

**Comment by galsalomon66** [comment]
please add compile-message to code, to make sure what type of regex-engine is actually been compiled.

File: CMakeLists.txt (hunk 14 above)
Comment (grajoria) [reply CMakeLists.txt]: Done

**Comment by grajoria** [comment]
>

File: CMakeLists.txt [hunk 17]
```diff
@@ -21,6 +21,32 @@ if(DEFINED ENV{NANO_SEC})
   add_definitions(-DBOOST_DATE_TIME_POSIX_TIME_STD_CONFIG)
 endif()
 
+
+set(REGEX_EN)
+if(REGEX_EN MATCHES "HS")
+  find_path(HYPERSCAN_INCLUDE_DIR NAMES hs/hs.h)
+  find_library(HYPERSCAN_LIB1 NAMES hs)
+  find_library(HYPERSCAN_LIB2 NAMES hs_runtime)
```
Comment (galsalomon66) [inline CMakeLists.txt:29]: why not using find_package( ... QUIET ) --> if ( ...FOUND)

File: include/s3select_oper.h [hunk 18]
```diff
@@ -15,6 +15,14 @@
 #include <boost/bind.hpp>
 #include "s3select_parquet_intrf.h" //NOTE: should include first (c++11 std::string_view)
 
+#if REGEX_HS
```
Comment (galsalomon66) [inline include/s3select_oper.h:18]: s3select CMake is not used by RGW CMake files.
the RGW is using only the s3select header files.

the header file can validate whether RE2/HS exists, by using the `__has_include (<....>)`  operator.
[https://github.com/ceph/s3select/blob/1609bb2ab5441d2314f56858f4f98fc2be509f89/include/s3select_parquet_intrf.h#L4](url)
 


the `REGEX_EN` definition can be used internally by s3select CMAKE, for selecting regex engine (for testing)

File: CMakeLists.txt [hunk 19]
```diff
@@ -21,6 +21,32 @@ if(DEFINED ENV{NANO_SEC})
...
+  set(HYPERSCAN_LIB $(HYPERSCAN_LIB1) $(HYPERSCAN_LIB2))
+  if(HYPERSCAN_INCLUDE_DIR AND HYPERSCAN_LIB)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -lhs")
```
Comment (galsalomon66) [inline CMakeLists.txt:32]: the definition is not

File: CMakeLists.txt (hunk 17 above)
Comment (grajoria) [reply CMakeLists.txt:29]: find_package is not working for re2 and hyperscan in cmake so I used find_path and find_library

File: include/s3select_oper.h (hunk 18 above)
Comment (grajoria) [reply include/s3select_oper.h]: Done

File: include/s3select_oper.h [hunk 20]
```diff
@@ -15,6 +15,15 @@
 #include <boost/bind.hpp>
 #include "s3select_parquet_intrf.h" //NOTE: should include first (c++11 std::string_view)
 
+
+#if __has_include (<hs/hs.h>) && REGEX_HS
```
Comment (galsalomon66) [inline include/s3select_oper.h:19]: :+1:

File: include/s3select_oper.h [hunk 21]
```diff
@@ -2676,6 +2685,387 @@ class base_timestamp_to_string : public base_function
...
 };
 
+
+class base_like : public base_function
+{
+  protected:
+    value like_expr_val;
+    value escape_expr_val;
+    bool constant_state = false;
+    #if REGEX_HS
```
Comment (galsalomon66) [inline include/s3select_oper.h:2695]: upon `REGEX_HS` is set 
but HS is not installed (headers are missing)
Does the build fail?
same question for RE2

**Comment by galsalomon66** [comment]
good work.
Thanks.
